### ISSUE TITLE
Extending `RenameProvider`

### ISSUE THREAD

**Issue by LoZander** [flix/flix#9561]
Currently, the `RenameProvider` is very limited due limitations on `SymUse`. This basically comes down to `SymUse`s including the entire qualified name in the `SourceLocation`.

**Comment by LoZander** [comment]
For full rename support, it would require some fairly deep changes I think. We'd need to know what ever part of the qualified name is a reference to. Maybe make `NName` into a list of `SymUse` or something. At least at some point, we'd need to create a list of `SymUse` or something like it for each part of the `NName`. This way we'd be able to rename not just things that might appear at the end of a qualified name, but also things that may appear at the start or in the middle of one, like modules, enums, effects, etc.

I think this change might be too big right now though. So I suggest a smaller change. Let's just change the `SouceLocations`s of `SymUse`s to only include the actual symbol and not the entire qualified name. This way we'd get the ability to rename things that may appear at the every end of a qualified name. So we'd get
- Functions
- Enum cases
- Signatures
- Effect ops
- Structs
- Type aliases

We'd still be missing
- Modules
- Enums
- Traits
- Effects

The most important thing, in my opinion, is that we get the ability to rename functions, as I reckon these are what a user would want to rename most often.

Any thoughts on this?

**Comment by magnus-madsen** [comment]
DefnSym is harder to change than I think e.g. EffSym hence the suggestion to start there.

**Comment by magnus-madsen** [comment]
I would go for a minimal PR that extends EffSym with a QName and uses that in the rename provider.

**Comment by LoZander** [comment]
But wouldn't we get a lot by just changing the `SourceLocations` like I suggest?

**Comment by magnus-madsen** [comment]
> But wouldn't we get a lot by just changing the `SourceLocations` like I suggest?

How is that not a change with huge implications? Why is it "the right thing to do"?

**Comment by LoZander** [comment]
Well, my thought was that the "symbol use" is not really the entire qualified name. The symbol is only a part of it. But yeah, changing the sourcelocations for `SymUses` is not a small change. Wouldn't using the QName in the provider to manually adjust sourcelocations be a bit of a hack? Wouldn't that fall outside the responsibility of the provider?

**Comment by LoZander** [comment]
The problem is this. Even with the QName, that doesn't give us any information on the SourceLocation of the use site of the symbol at all, only the declaration site. We get the namespace and identifier of the symbol, but we more or less already have that. So I don't see how it actually helps us. The SymUses already contain the symbols which contain the namespace (as a list of strings). Having the QName, I don't see how that gives us more useful information to work with.

What we need is SourceLocation information on the qualified name at the use site. For `Something.DivByZero.divByZero` we need source location information on the `DivByZero` part. Adding QName doesn't get us that. We'd still have to manually extract that part from the SymUse SourceLocation, which is the exact kind of hack we wanted to avoid

**Comment by magnus-madsen** [comment]
What if the SymUse contains a qname?

**Comment by LoZander** [comment]
I don't think it would make a difference, at least not if I've understood QName correctly. I believe that QName refers to the qualified name of the symbol at the declaration site. That's not particularly useful, at least not more than the namespace that we already have access to through the symbol itself. We can ofc. always use it to figure out how much we need to chop off of the SourceLocation, but then we're back to hack job.

The problem is that our current model of SymUse sees something like `Math.Div.DivByZero.divByZero` like a single SymUse, while it's actually using four symbols. At no point, as far as I know, do we record the SourceLocations for the occurrences `Math` or `Div`, `DivByZero` or `divByZero` in that qualified name. Having the QName could tell us that `divByZero` is placed in `Math.Div.DivByZero`, but it doesn't tell us the SourceLocation of the occurrences of these other symbols in that string of symbols. WIth the QName, the best we could hope to do is just leveraging it to figure out how to cut up single SourceLocation for SymUse that we do have, since we know this one covers the entire thing. But I think it's a mistake that this covers the entire thing to begin with. And if we do take that appraoch, then we'll run into issues when it comes to imports and such. And tghe information in QName that might be of interest in this approach is something we already have available as the `namespace` member of the symbol.

**Comment by LoZander** [comment]
The QName is better than the `namespace` member we have access to, because it has the namespace as a list of idents as opposed to a list of strings. So we could use it to figure out what symbols are in the SymUse by just disecting SourceLocation associated with the SymUse. We could bundle the QName in the SymUse and do this at the provider, or do this when constructing the SymUse and extend the SymUse with a list of SymUses for the qualified name occurrence.

**Comment by LoZander** [comment]
Maybe something like that might be a way to proceed?

**Comment by LoZander** [comment]
Expend SymUse with a NamespaceUse or something like that?

**Comment by magnus-madsen** [comment]
A QName contains the source location of its occurrence not of what it points to.

**Comment by LoZander** [comment]
Oh, then I misunderstood. I'm very happy to have been mistaken because that makes things much simpler. Then yes, absolutely, putting a QName in SymUse solves the issue :D

**Comment by LoZander** [comment]
To rename effects, we will also need to extend `SigSymUse` with their QName, since we need to rename the occurrence of the effect name in SigSymUse's qualified name

**Comment by magnus-madsen** [comment]
> To rename effects, we will also need to extend `SigSymUse` with their QName, since we need to rename the occurrence of the effect name in SigSymUse's qualified name. I'll do that next. Then renaming effects should be easy :)

I don't understand what effects have to do with signatures.

**Comment by LoZander** [comment]
> > To rename effects, we will also need to extend `SigSymUse` with their QName, since we need to rename the occurrence of the effect name in SigSymUse's qualified name. I'll do that next. Then renaming effects should be easy :)
> 
> I don't understand what effects have to do with signatures.

Oh, ofc. not. They have nothing to do with each other. I meant `OpSimUse` :p
